3.7 alpha1 does not compile on earlier FreeBSD releases

syslog-ng 3.7 alpha1 compiles fine on FreeBSD 10.1 (the latest release), but fails to compile on earlier releases. Originally reported by Cy (FreeBSD ports maintainer) on FreeBSD 8.4 and 9.1. Reproduced here on FreeBSD 9.2:

root@fb92:/usr/ports/sysutils/syslog-ng-devel # make
===>  Building for syslog-ng-devel-3.7.0a1
gmake --no-print-directory all-recursive
  CC       lib/lib_libsyslog_ng_la-logmsg.lo
In file included from lib/logmsg.c:37:
lib/template/macros.h:30: error: redefinition of typedef 'LogTemplateOptions'
lib/template/templates.h:89: error: previous declaration of 'LogTemplateOptions' was here
Makefile:6851: recipe for target 'lib/lib_libsyslog_ng_la-logmsg.lo' failed
gmake[2]: *** [lib/lib_libsyslog_ng_la-logmsg.lo] Error 1
Makefile:9939: recipe for target 'all-recursive' failed
gmake[1]: *** [all-recursive] Error 1
Makefile:4068: recipe for target 'all' failed
gmake: *** [all] Error 2
*** [do-build] Error code 1

Stop in /usr/ports/sysutils/syslog-ng-devel.
*** [stage] Error code 1

Stop in /usr/ports/sysutils/syslog-ng-devel.
root@fb92:/usr/ports/sysutils/syslog-ng-devel #

--- a/lib/template/common-template-typedefs.h
+++ b/lib/template/common-template-typedefs.h
@@ -1,0 +1,29 @@
+/*
+ * Copyright (c) 2014 BalaBit S.a.r.l., Luxembourg, Luxembourg
+ * Copyright (c) 2014 Laszlo Budai
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#ifndef COMMON_TYPEDEFS_H_INCLUDED
+#define COMNON_TYPEDEFS_H_INCLUDED
+
+typedef struct _LogTemplateOptions LogTemplateOptions;
+
+#endif

--- a/lib/template/macros.h
+++ b/lib/template/macros.h
@@ -26,8 +26,7 @@
 #define TEMPLATE_MACROS_H_INCLUDED
 
 #include "syslog-ng.h"
-
-typedef struct _LogTemplateOptions LogTemplateOptions;
+#include "common-template-typedefs.h"
 
 /* macro IDs */
 enum

--- a/lib/template/templates.h
+++ b/lib/template/templates.h
@@ -26,6 +26,7 @@
 #define TEMPLATES_H_INCLUDED
 
 #include "syslog-ng.h"
+#include "common-template-typedefs.h"
 #include "timeutils.h"
 #include "type-hinting.h"
 #include "plugin.h"
@@ -72,7 +73,7 @@ typedef struct _LogTemplate
  * is static throughout the runtime for a given configuration. There
  * are call-site specific options too, those are specified as
  * arguments to log_template_format() */
-typedef struct _LogTemplateOptions
+struct _LogTemplateOptions
 {
   gboolean initialized;
   /* timestamp format as specified by ts_format() */
@@ -86,7 +87,7 @@ typedef struct _LogTemplateOptions
 
   /* Template error handling settings */
   gint on_error;
-} LogTemplateOptions;
+};
 
 /* appends the formatted output into result */
 

